pkgs/by-name/in/install-grub-ng/src/builder.rs

```diff
@@ -0,0 +1,153 @@
+mod appearance;
+mod entries;
+mod install;
+
+use std::{
+    collections::HashSet,
+    fmt::Write as _,
+    fs,
+    os::{linux::fs::MetadataExt, unix::fs::PermissionsExt},
+    path::{Path, PathBuf},
+};
+
+use eyre::Result;
+
+use crate::{
+    config::{Config, Password},
+    grub::Grub,
+};
+
+pub struct Builder<'conf> {
+    config: Config<'conf>,
+
+    grub_boot: Grub,
+    grub_store: Option<Grub>,
+
+    default_config: &'conf Path,
+    pub copied: HashSet<PathBuf>,
+
+    dry_run: bool,
+
+    inner: String,
+}
+impl<'conf> Builder<'conf> {
+    pub fn new(mut config: Config<'conf>, default_config: &'conf Path) -> Result<Self> {
+        let dry_run = std::env::var("DRY_RUN").as_deref() == Ok("true");
+
+        if !dry_run {
+            let grub = config.boot_path.join("grub");
+            fs::create_dir_all(&grub)?;
+            fs::set_permissions(&grub, fs::Permissions::from_mode(0o700))?;
+        }
+
+        // Discover whether the bootPath is on the same filesystem as / and
+        // /nix/store.  If not, then all kernels and initrds must be copied
+        // to the bootPath.
+        if let Ok(metadata) = config.boot_path.metadata() {
+            if metadata.st_dev() != Path::new("/nix/store").metadata()?.st_dev() {
```

Comment:
The nix store can technically be at a different path, ideally it would be passed in via the config using `builtins.storeDir`.
Just a TODO would be ok for now.